Use `atexit to` clean up after simulators

mundya opened the pull request:
`Simulator.close` is registered as an `atexit` method to help in cases such as people Ctrl+C-ing the GUI.

mundya committed:
Use atexit to clean up after simulators

`close` is registered as an `atexit` method to help in cases such as
people Ctrl+C-ing the GUI.

tcstewar commented on nengo_spinnaker/simulator.py:
```diff
@@ -47,6 +48,9 @@ def __init__(self, network, dt=0.001, period=10.0):
             memory will be allocated to store precomputed and probed data.
         """
+        # Register `close` to be called when the interpreter is closed.
+        atexit.register(self.close)
```
Registering this close function will stop Python from ever actually deleting any Simulator object, since there's always going to be a reference to it, even if it has been manually closed.  It might be better to just add one static atexit handler and keep a static list of unclosed Simulators.

mundya replied:
Nice, I hadn't thought of that.  This might be a job for `weakref`.

tcstewar commented on nengo_spinnaker/simulator.py:
```diff
@@ -278,10 +290,18 @@ def func(t, x, f=old_func):
+    """Calls close on a simulator if it still exists."""
+    obj = ref()
+    if obj is not None:
+        obj.close()
```
Hmm...  I might be misreading what's happening here, but it looks to me like if I create a Simulator and then forget my reference to it without closing, then it'll never get closed.

mundya replied:
I'm not doing well on this task today. You're right, I shall fix that!

mundya committed:
Simulator keeps track of open simulators

 - On creation a simulator is added to a list of open simulators
 - On closing a simulator it is removed from the list of open
   simulators
 - On exit all open simulators are closed

tcstewar commented:
Looks good to me.
